ui/src/coordinator.rs

```diff
@@ -0,0 +1,1284 @@
+        let WSFormatRequest {
+            code,
+            edition,
+            extra,
+        } = value;
+        let request = sandbox::FormatRequest {
+            code,
+            edition: parse_edition(&edition)?,
+        };
+        Ok((request, extra))
+    }
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WSFormatResponse {
+    success: bool,
+    code: String,
+    stdout: String,
+    stderr: String,
+    extra: serde_json::Value,
+}
+
+impl From<(sandbox::FormatResponse, serde_json::Value)> for WSFormatResponse {
+    fn from(value: (sandbox::FormatResponse, serde_json::Value)) -> Self {
+        let sandbox::FormatResponse {
+            success,
+            code,
+            stdout,
+            stderr,
+        } = value.0;
+        WSFormatResponse {
+            success,
+            code,
+            stdout,
+            stderr,
+            extra: value.1,
+        }
+    }
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WSExecuteRequest {
+    channel: String,
+    mode: String,
+    edition: String,
+    crate_type: String,
+    tests: bool,
+    code: String,
+    backtrace: bool,
+    extra: serde_json::Value,
+}
+
+impl TryFrom<WSExecuteRequest> for (sandbox::ExecuteRequest, serde_json::Value) {
+    type Error = crate::Error;
+
+    fn try_from(value: WSExecuteRequest) -> Result<Self, Self::Error> {
+        let WSExecuteRequest {
+            channel,
+            mode,
+            edition,
+            crate_type,
+            tests,
+            code,
+            backtrace,
+            extra,
+        } = value;
+
+        let req = sandbox::ExecuteRequest {
+            channel: parse_channel(&channel)?,
+            mode: parse_mode(&mode)?,
+            edition: parse_edition(&edition)?,
+            crate_type: parse_crate_type(&crate_type)?,
+            tests,
+            backtrace,
+            code,
+        };
+
+        Ok((req, extra))
+    }
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct WSExecuteResponse {
+    success: bool,
+    stdout: String,
+    stderr: String,
+    extra: serde_json::Value,
+}
+
+impl From<(sandbox::ExecuteResponse, serde_json::Value)> for WSExecuteResponse {
+    fn from(value: (sandbox::ExecuteResponse, serde_json::Value)) -> Self {
+        let sandbox::ExecuteResponse {
+            success,
+            stdout,
+            stderr,
+        } = value.0;
+        let extra = value.1;
+
+        WSExecuteResponse {
+            success,
+            stdout,
+            stderr,
+            extra,
+        }
+    }
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+enum WSMessage {
+    Request(WSRequest),
+    StdinPacket(String),
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(tag = "type")]
+enum WSRequest {
```

Comment:
`ExecuteRequest` has different behavior from other two as it enables streaming.

But yes, we can abandon `FormatRequest` for now.

Reply:
Yep. I often advocate for a [*walking skeleton*](https://wiki.c2.com/?WalkingSkeleton) that focuses on one of the tougher aspects of the system. Streaming is new and unique to the playground, which is why it would have been my choice.

Let's move forward with compilation, not formatting or execution. You've mentioned MIR a few times, but is your expectation that other things like LLVM / assembly / Wasm are also all supported?

Reply:
In current design, streaming is tied to `cargo run` which is `ExecuteRequest`. 
`CompileRequest` doesn't get streaming. 
Do you mean we support streaming `cargo build/rustc' too?

MIR is an example of `CompileRequest`. But as it concerns frontend code change, only MIR can be seen by normal users. All LLVM/assembly/Wasm `CompileRequest` should work in the backend and also can be seen from browser developer tool.

Reply:
Yes, `cargo run` is going to be the primary use case for streaming I/O (testing will likely also want it eventually).

For this first PR, since we are focusing on `CompileRequest`, I do not think we need to worry about any kind of I/O streaming[^1].

> only MIR can be seen by normal users

Ok, I can do a follow up PR to enable it for the others in the UI. I actually expect that it will *not* work for Wasm in the backend right away, as we have to do some ugly things to get that to work.

[^1]: In the future, I might experiment with splitting `cargo build` and `cargo run` into separate parts so that the UI can show them distinctly, and as part of that I might also stream the data, but that's not important for now.